File: BaekjoonOnlineJudge/DFS/단지번호붙이기 (2667)/Solution.py
n = int(input())

graph = []
for i in range(n):
  graph.append(list(map(int, input())))


def dfs(x, y, cnt):
  if x <= -1 or x >= n or y <= -1 or y >= n:
    return False
  if graph[x][y] == 1:
    graph[x][y] = 0
    cnt += 1
    cnt += dfs(x-1, y, 0)
    cnt += dfs(x, y-1, 0)
    cnt += dfs(x+1, y, 0)
    cnt += dfs(x, y+1, 0)
    return cnt
  return cnt


total_village = 0
villages = []

for i in range(n):
  for j in range(n):
    result = dfs(i, j, 0)
    if result != 0:
      total_village += 1
      villages.append(result)

print(total_village)
for village in sorted(villages):
  print(village)

File: BaekjoonOnlineJudge/DFS/단지번호붙이기 (2667)/Solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int n;
int graph[25][25];

int dfs(int a,int b) {
	int count = 0;
	if(a<=-1||a>=n||b<=-1||b>=n||graph[a][b]==0)
		return 0;

	if (graph[a][b] == 1) {
		count++;
		graph[a][b] = 0;

		count += dfs(a - 1, b);
		count += dfs(a, b - 1);
		count += dfs(a + 1, b);
		count += dfs(a, b + 1);
		return count;
	}

	return 0;
}

int main() {
	cin >> n;
	vector<int> result;

	for (int i = 0;i < n;i++) {
		for (int j = 0;j < n;j++)
			scanf("%1d", &graph[i][j]);
	}

	for (int i = 0;i < n;i++) {
		for (int j = 0;j < n;j++) {
			int count = dfs(i, j);
			if (count != 0) {
				result.push_back(count);
			}
		}
	}

	sort(result.begin(), result.end());
	cout << result.size() << endl;

	for (int i = 0;i < result.size();i++)
		cout << result[i] << endl;
}
